onnx/optimizer/passes/eliminate_unused_initializer.h

```diff
@@ -0,0 +1,80 @@
+// ATTENTION: The code in this file is highly EXPERIMENTAL.
+// Adventurous users should note that the APIs will probably change.
+
+#pragma once
+
+#include "onnx/optimizer/passes/optimize_pass.h"
+
+namespace ONNX_NAMESPACE { namespace optimization {
+
+struct EliminateUnusedInitializer final : public OptimizePass {
+  explicit EliminateUnusedInitializer()
+    : OptimizePass("eliminate_unused_initializer", API_TYPE::IR) {
+  }
+
+  void eliminate_unused_initializer(Graph& graph) {
+    std::vector<Tensor> new_initializers;
+    std::vector<std::string> new_initializer_names;
+    for (auto it = graph.begin(); it != graph.end(); ++it) {
+      auto* n = *it;
+      DescendOnGraphAttributes(n, [this](Graph& g){eliminate_unused_initializer(g);});
+      // put any initializer used as input of any node to new vector
+      for (auto* input : n->inputs()) {
+        if (std::find(
+                graph.initializer_names().begin(),
+                graph.initializer_names().end(),
+                input->uniqueName()) != graph.initializer_names().end()) {
+          Tensor initializer;
+          for (auto _init : graph.initializers()) {
+            if (_init.name() == input->uniqueName()) {
+              initializer = _init;
+            }
+          }
+          new_initializers.push_back(initializer);
+          new_initializer_names.push_back(input->uniqueName());
+        }
+      }
+    }
+    // get diff between old and new initializers vector
+    std::vector<std::string> diff_initializer_names, removed_initializer_names;
+    std::set_difference(
```

Comment:
it assumes sorted ranges which is not the case. Do you want to use std::unordered_set instead?